## Q-table in the case of Frozen Lake (Slippery is True) problem

In [ ]:
import gymnasium as gym
import time
import numpy as np
import matplotlib.pyplot as plt
env = gym.make("FrozenLake-v1", is_slippery=True, render_mode="ansi")
env.reset()
action_size = env.action_space.n
print("Action size: ", action_size)
state_size = env.observation_space.n
print("State size: ", state_size)
print(env.render())


def eval_policy_better(env_, pi_, gamma_, t_max_, episodes_):
    
    goal_reached_number = 0 #to_count_how_many_times_the_agent_reached_the_goal
    
    v_pi_rep = np.empty(episodes_) # N trials
    for e in range(episodes_):
        s_t = env.reset()[0]
        v_pi = 0
        for t in range(t_max_):
            a_t = pi_[s_t]
            s_t, r_t, done, truncated, info = env_.step(a_t) 
            v_pi += gamma_**t*r_t
            if done:
                if r_t == 1:
                    goal_reached_number = goal_reached_number+1
                break
        v_pi_rep[e] = v_pi
        env.close()
    print(f'The agent reached its goal {goal_reached_number} times.')
    return np.mean(v_pi_rep), np.min(v_pi_rep), np.max(v_pi_rep), np.std(v_pi_rep)

ModuleNotFoundError: No module named 'gymnasium'

In [1]:
import sys; print(sys.executable); import site; print(site.getsitepackages())

c:\Users\Lenovo Legion\Desktop\intro_ml\mnist_noise\.venv\Scripts\python.exe
['c:\\Users\\Lenovo Legion\\Desktop\\intro_ml\\mnist_noise\\.venv', 'c:\\Users\\Lenovo Legion\\Desktop\\intro_ml\\mnist_noise\\.venv\\Lib\\site-packages']


Q-table by Mihail. I have taken the 100000 episodes so that the Value function mean and reached_goals_number don't deviate much in every run.

In [10]:
gamma = 0.9
Q = np.zeros([state_size, action_size])
Q[0,0] = 1
Q[1,3] = 1
Q[2,3] = 1
Q[3,3] = 1
Q[4,0] = 1
Q[6,1] = 1
Q[8,2] = 1
Q[9,1] = 1
Q[10,0] = 1
Q[13,2] = 1
Q[14,1] = 1 

print(Q)

pi_Q = np.argmax(Q,axis=1)
print(pi_Q)
val_mean, val_min, val_max, val_std = eval_policy_better(env, pi_Q, gamma, 666, 100000)
print(f'Value function mean {val_mean:.4f}, min {val_min:.4f} max {val_max:.4f} and std {val_std:.4f}')

[[1. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 1. 0. 0.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 1. 0. 0.]
 [0. 0. 0. 0.]]
[0 3 3 3 0 0 1 0 2 1 0 0 0 2 1 0]
The agent reached its goal 21903 times.
Value function mean 0.0380, min 0.0000 max 0.5905 and std 0.0998


Well, the Value function mean is 0.0380, and the agent reached the goal 21903 times successfully. That means only 21.86% times it raches its goal. 

I will analyze two separate Q-tables that I have deduced and compare their results later. First, I will change the actions in state 6 and state 8 in Mihail's Q-table. 

In [20]:
Q = np.zeros([state_size, action_size])
Q[0,0] = 1
Q[1,3] = 1
Q[2,3] = 1
Q[3,3] = 1
Q[4,0] = 1
Q[6,2] = 1   #I chose action 2 (go right which means towards the lake instead of going down)
Q[8,3] = 1   #I chose action 3 (go up which instead of going right), this action is the major development in this Q-table
Q[9,1] = 1
Q[10,0] = 1
Q[13,2] = 1
Q[14,1] = 1 

print(Q)

pi_Q = np.argmax(Q,axis=1)
print(pi_Q)
val_mean, val_min, val_max, val_std = eval_policy_better(env, pi_Q, gamma, 666, 100000)
print(f'Value function mean {val_mean:.4f}, min {val_min:.4f} max {val_max:.4f} and std {val_std:.4f}')

[[1. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 1. 0. 0.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 1. 0. 0.]
 [0. 0. 0. 0.]]
[0 3 3 3 0 0 2 0 3 1 0 0 0 2 1 0]
The agent reached its goal 82234 times.
Value function mean 0.0679, min 0.0000 max 0.5905 and std 0.1129


Aha! The Value function mean is now 0.0679 (much better), and the agent successfully reached the goal 82234 times (3.75 times than the previous one). That means 82.25% times it raches its goal. A good development. 

In this step, I chose to jump into the lake instead of trying to avoid it. Which is probabilistically more preferred anyway (more in my PDF). 

**That’s life**; It's true that you need to go towards your goal, avoiding problems. Sometimes you need to do some crazy stuff like jumping upward or toward the wall. But now we know, we need to take risks also, like jumping to the lake where death is inevitable. It seems scary. Your brain will try to stop this action. But you need to take risks to do better in your life. But my suggestion is to take calculated risk just like my Qtable.  

Now, I will change the action in state 2 in my latest Q-table and will observe what will happen. 

In [21]:
Q = np.zeros([state_size, action_size])
Q[0,0] = 1
Q[1,3] = 1
Q[2,1] = 1
Q[3,3] = 1
Q[4,0] = 1
Q[6,2] = 1
Q[8,3] = 1
Q[9,1] = 1
Q[10,0] = 1
Q[13,2] = 1
Q[14,1] = 1 

print(Q)

pi_Q = np.argmax(Q,axis=1)
print(pi_Q)
val_mean, val_min, val_max, val_std = eval_policy_better(env, pi_Q, gamma, 666, 100000)
print(f'Value function mean {val_mean:.4f}, min {val_min:.4f} max {val_max:.4f} and std {val_std:.4f}')

[[1. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 1. 0. 0.]
 [0. 0. 0. 1.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 1.]
 [0. 1. 0. 0.]
 [1. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 1. 0.]
 [0. 1. 0. 0.]
 [0. 0. 0. 0.]]
[0 3 1 3 0 0 2 0 3 1 0 0 0 2 1 0]
The agent reached its goal 78281 times.
Value function mean 0.0693, min 0.0000 max 0.5905 and std 0.1142


Look! The Value function mean is now 0.0693 (slightly better total reward points achieved than my previous Q-table). However, the agent successfully reached the goal 78281 times (less than the previous one). So, the previous Qtable was better for reaching the goal successfully, but this one is better in the case of collecting rewards. I will discuss all these issues more in my PDF (Observations and reasoning behind my Q-table in the case of Frozen Lake (Slippery is True) problem).